### PR TITLE
LOOP-4752 Integrate stateless algorithm into Loop

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-4752
Changes for Loop to use stateless/functional version of Loop Algorithm. Additional changes:
* Therapy settings largely removed from data stores
* Algorithm related math removed from data stores.
* Wrappers added to stores to use async/await

### CODE REVIEW

File: LoopKit/GlucoseKit/GlucoseStore.swift [hunk 1]
```diff
@@ -703,61 +712,6 @@ extension GlucoseStore {
     }
 }
 
-// MARK: - Math
```
Comment (ps2) [inline LoopKit/GlucoseKit/GlucoseStore.swift:706]: Moving diabetes math out of the stores; now part of the Loop algorithm.

File: LoopKit/CarbKit/CarbMath.swift [hunk 2]
```diff
@@ -16,7 +16,21 @@ public struct CarbMath {
     public static let defaultEffectDelay: TimeInterval = .minutes(10)
 }
 
-struct CarbModelSettings {
+public enum CarbAbsorptionModel {
```
Comment (ps2) [inline LoopKit/CarbKit/CarbMath.swift:19]: Moved from CarbStore; stores are no longer responsible for algorithm math.

File: LoopKit/GlucoseKit/GlucoseStore.swift [hunk 3]
```diff
@@ -947,32 +901,34 @@ extension GlucoseStore {
     /// This operation is performed asynchronously and the completion will be executed on an arbitrary background queue.
     ///
     /// - parameter completionHandler: A closure called once the report has been generated. The closure takes a single argument of the report string.
-    public func generateDiagnosticReport(_ completionHandler: @escaping (_ report: String) -> Void) {
```
Comment (ps2) [inline LoopKit/GlucoseKit/GlucoseStore.swift:950]: Loop issue report everywhere has been converted to async.

File: LoopKit/InsulinKit/AutomaticDoseRecommendation.swift [hunk 4]
```diff
@@ -9,13 +9,17 @@
 import Foundation
 
 public struct AutomaticDoseRecommendation: Equatable {
-    public let basalAdjustment: TempBasalRecommendation?
-    public let bolusUnits: Double?
-    
+    public var basalAdjustment: TempBasalRecommendation?
+    public var bolusUnits: Double?
```
Comment (Camji55) [inline LoopKit/InsulinKit/AutomaticDoseRecommendation.swift:13]: Do these need to be a `var`?

File: LoopKit/InsulinKit/TempBasalRecommendation.swift [hunk 5]
```diff
@@ -10,8 +10,8 @@ import Foundation
 import HealthKit
 
 public struct TempBasalRecommendation: Equatable {
-    public let unitsPerHour: Double
-    public let duration: TimeInterval
+    public var unitsPerHour: Double
+    public var duration: TimeInterval
```
Comment (Camji55) [inline LoopKit/InsulinKit/TempBasalRecommendation.swift]: Are these ever mutated?

**Review by Camji55 — APPROVED** [review]
LGTM. It would be really great if the Loop Algo didn't have a dependency on HealthKit (it seems this is only to use the `HKQuantity` type?) and instead we had a wrapper around the HK models. That in theory should allow it to run on machines that don't have HK which might be useful?

**Comment by ps2** [comment]
> LGTM. It would be really great if the Loop Algo didn't have a dependency on HealthKit (it seems this is only to use the `HKQuantity` type?) and instead we had a wrapper around the HK models. That in theory should allow it to run on machines that don't have HK which might be useful?

Yes, I agree 100%. I also want to make the other input types for insulin, carbs, and glucose from concrete types into protocols, so we can be more specific on what's required, and flexible about what we accept. I tried that initially, and then backed off, as it was causing too much code explosion, but I do think it should be possible.

File: LoopKit/InsulinKit/AutomaticDoseRecommendation.swift (hunk 4 above)
Comment (ps2) [reply LoopKit/InsulinKit/AutomaticDoseRecommendation.swift:13]: LoopDataManager does adjust these to round them to deliverable increments.

File: LoopKit/InsulinKit/TempBasalRecommendation.swift (hunk 5 above)
Comment (ps2) [reply LoopKit/InsulinKit/TempBasalRecommendation.swift]: unitsPerHour is modified during rounding to deliverable increments, but duration is not; I'll change it to a let.

File: LoopKit/DeviceManager/PumpManager.swift [hunk 6]
```diff
@@ -251,3 +251,50 @@ public extension PumpManager {
         }
     }
 }
+
+public extension PumpManager {
+
```
Comment (nhamming) [inline LoopKit/DeviceManager/PumpManager.swift:256]: should this include suspend and resume?

**Review by nhamming — APPROVED** [review]
LGTM

File: LoopKit/DeviceManager/PumpManager.swift (hunk 6 above)
Comment (ps2) [reply LoopKit/DeviceManager/PumpManager.swift:256]: So far, suspend/resume is called by PumpManager UI, and not by Loop; I think even the non-async methods might be able to be removed?

**Pull request merged**